pkg/storegateway/indexheader/reader_pool_test.go

```diff
@@ -138,3 +125,100 @@ func TestReaderPool_ShouldCloseIdleLazyReaders(t *testing.T) {
 	require.Equal(t, float64(2), promtestutil.ToFloat64(metrics.lazyReader.loadCount))
 	require.Equal(t, float64(2), promtestutil.ToFloat64(metrics.lazyReader.unloadCount))
 }
+
+func TestReaderPool_LoadedBlock(t *testing.T) {
+	usedAt := time.Now()
+	id, err := ulid.New(ulid.Now(), rand.Reader)
+	require.NoError(t, err)
+
+	lb := LazyBinaryReader{
+		blockID: id,
+		usedAt:  atomic.NewInt64(usedAt.UnixNano()),
+		// we just set to make reader != nil
+		reader: &StreamBinaryReader{},
+	}
+	rp := ReaderPool{
+		lazyReaderEnabled: true,
+		lazyReaders:       map[*LazyBinaryReader]struct{}{&lb: {}},
+	}
+	require.Equal(t, map[ulid.ULID]int64{id: usedAt.UnixMilli()}, rp.LoadedBlocks())
+}
+
+func TestReaderPool_PersistLazyLoadedBlock(t *testing.T) {
+	const idleTimeout = time.Second
+	ctx, tmpDir, bkt, blockID, metrics := prepareReaderPool(t)
+
+	// Note that we are creating a ReaderPool that doesn't run a background cleanup task for idle
+	// Reader instances. We'll manually invoke the cleanup task when we need it as part of this test.
+	pool := newReaderPool(log.NewNopLogger(), true, idleTimeout, metrics)
+	defer pool.Close()
+
+	r, err := pool.NewBinaryReader(ctx, log.NewNopLogger(), bkt, tmpDir, blockID, 3, Config{})
+	require.NoError(t, err)
+	defer func() { require.NoError(t, r.Close()) }()
+
+	// Ensure it can read data.
+	labelNames, err := r.LabelNames()
+	require.NoError(t, err)
+	require.Equal(t, []string{"a"}, labelNames)
+	require.Equal(t, float64(1), promtestutil.ToFloat64(metrics.lazyReader.loadCount))
+	require.Equal(t, float64(0), promtestutil.ToFloat64(metrics.lazyReader.unloadCount))
+
+	snapshot := lazyLoadedHeadersSnapshot{
+		HeaderLastUsedTime: pool.LoadedBlocks(),
+		UserID:             "anonymous",
+	}
+
+	err = snapshot.persist(tmpDir)
+	require.NoError(t, err)
+
+	persistedFile := filepath.Join(tmpDir, lazyLoadedHeadersListFile)
+	persistedData, err := os.ReadFile(persistedFile)
+	require.NoError(t, err)
+
+	var expected string
+	// we know that there is only one lazyReader, hence just use formatter to set the ULID and timestamp.
```

Comment:
If we expect that there's only one lazy reader, how about we do something like this?
Then if that expectation is ever broken, the test will break in a predictable way.

```go
require.Len(t, 1, pool.lazyReaders)
reader := pool.lazyReaders[0]
expected := fmt.Sprintf(`{"header_last_used_time":{"%s":%d},"user_id":"anonymous"}`, reader.blockID, reader.usedAt.Load()/int64(time.Millisecond))
```

Reply:
I added assertion to make sure only one lazyReader. But since lazyReaders is a map, I don't find an easy way to get the first/only element. That was the reason I used for loop.

Reply:
Right, makes sense, missed that.